cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -460,9 +476,53 @@ func (m *McmManager) GetMachineDeploymentNodeTemplate(machinedeployment *Machine
 		}
 		region = mc.Spec.Location
 		if mc.Spec.Properties.Zone != nil {
-			// Do not re-use the zone before re-vendoring mcm.
 			zone = mc.Spec.Location + "-" + strconv.Itoa(*mc.Spec.Properties.Zone)
 		}
+	case kindMachineClass:
+		mc, err := m.machineclient.MachineClasses(m.namespace).Get(context.TODO(), machineClass.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("Unable to fetch %s for %s, Error: %v", kindMachineClass, machinedeployment.Name, err)
+		}
+		switch mc.Provider {
+		case providerAWS:
+			var providerSpec *awsapis.AWSProviderSpec
+			err = json.Unmarshal(mc.ProviderSpec.Raw, &providerSpec)
+			if err != nil {
+				return nil, fmt.Errorf("Unable to convert from %s to %s for %s, Error: %v", kindMachineClass, providerAWS, machinedeployment.Name, err)
+			}
+
+			awsInstance := aws.InstanceTypes[providerSpec.MachineType]
+			instance = instanceType{
+				InstanceType: awsInstance.InstanceType,
+				VCPU:         awsInstance.VCPU,
+				MemoryMb:     awsInstance.MemoryMb,
+				GPU:          awsInstance.GPU,
+			}
+			region = providerSpec.Region
+			if mc.Labels != nil {
```

Comment:
Here, we are reading the zone from the machineclass label. This is not a tight contract. How do we make sure the machineclasses are rightly labeled? Is this advisable?

Reply:
Hi @amshuman-kr ,

I think the current labelling is done from the [extension-provider-aws here](https://github.com/gardener/gardener-extension-provider-aws/pull/78/files#diff-ede719e7797a6dd04b91acb4296b78434730159686b71e616b1c989712a7716eR202). 

And like you have mentioned I don't think this is tight coupling currently.  On whether this is advisable? I think @hardikdr has more context to this. 

However, what we should definitely do is, change how we do scale to/from zero altogether and make it generic enough for all providers. As currently, it is still provider specific for the scale-to logic.

Reply:
I think it is ok for now. Later we should write a controller (a library to be used as part of the driver controller implementation) to label the machine class automatically and not via the worker controller or provider.

Reply:
Agreed to both points:
- We need to enable generic support for autoscaler.
- And also, there could be a dedicated field for the `zone` maybe. We don't have it in the AWSMachineClass already because MCM doesn't need it to create machines.

Reply:
Yes sure. Let's fix this in another round.